### DISCUSSION TITLE
Creating a flexible date-time input field

### DISCUSSION THREAD

**Discussion by nareshbhatia** [react-hook-form/react-hook-form#2108] [Q&A] [answered]
I am new to react-hook-form and looking for some guidance on how to implement a custom date-time input field. The field's type is a `Date` object, however it uses an input element to accept a date as a string. The idea is that the field should allow date input in a very flexible way, not using a strict mask. For example, user can type '2020-4-15' and the field should convert it to a proper `Date` object representing '2020-04-15 12:00 AM'. Underneath the scenes, I would like to use the flexible parsing and formatting features of [momentjs](https://momentjs.com/docs/#/parsing/string-format/).

The examples I saw in Getting Started are all dealing with strings (even the enum example uses strings). Is there any guidance on how to implement a field with more complex data types like `Date`?

FYI, I tried to implement this use case using formik and came very close (see my [CodeSandbox](https://codesandbox.io/s/formik-datetime-field-f3cn4?file=/src/DateTimeField.tsx)). However the field validation and resetForm() capabilities are broken. My only conclusion is that I cannot maintain an internal state in a field component that is different from the external state.

Wondering if react-hook-form provides an alternate approach.

**Comment by bluebill1049** [comment, marked as answer]
hey @nareshbhatia i think it will be easier to build a custom component (controlled) and let Controller do the value collection and validation job. you can find a similar example here: https://codesandbox.io/s/react-hook-form-v6-controller-qsd8r
